Lib/pathlib.py

```diff
@@ -61,6 +61,15 @@ def _is_case_sensitive(flavour):
 # Globbing helpers
 #
 
+
+_FNMATCH_PREFIX, _FNMATCH_SUFFIX = fnmatch.translate('_').split('_')
+_FNMATCH_SLICE = slice(len(_FNMATCH_PREFIX), -len(_FNMATCH_SUFFIX))
```

Comment:
This is a bit of a hack. Alternatively I could add an argument to `fnmatch.translate()` that makes it exclude the prefix/suffix, or add a new private `fnmatch._translate()` function perhaps? Not sure.

Reply:
@AlexWaygood any recommendations? Thanks

Reply:
Hrmmmmmmm. I don't mind a hack for an implementation detail, but this isn't particularly readable as-is :/

what about defining a new function here:

```suggestion
_FNMATCH_PREFIX, _FNMATCH_SUFFIX = fnmatch.translate('_').split('_')


def _fnmatch_translate_stripping_prefix(pat):
    """fnmatch.translate(), but with the prefix and suffix stripped from the regex
    
    >>> fnmatch.translate("_")
    '(?s:_)\\Z'
    >>> _fnmatch_translate_stripping_prefix("_")
    '_'
    """
    return fnmatch.translate(pat)[len(_FNMATCH_PREFIX) : -len(_FNMATCH_SUFFIX)]


```

And then you could use this new function on this line, where you currently do `fnmatch.translate(part)[_FNMATCH_SLICE]`. (The `len()` calls should take up almost no time at all, by the way, so I don't think there's any need for the `_FNMATCH_SLICE` variable, which makes it less readable imo.)

Reply:
I don't mind the current code _too_ much though. It took me a few seconds to grok what was going on, but I think it's perfectly fine to go ahead with if you don't like my suggestion :)

Reply:
Makes sense. Do you think there's any merit in moving that function to `fnmatch._translate_unwrapped()` or similar? `fnmatch.translate()` could be implemented on top, and fnmatch wouldn't add the prefix/suffix only for pathlib to strip it again. OTOH, pathlib would use a private function from another module, so probably not ideal?

Reply:
i.e.:

```diff
diff --git a/Lib/fnmatch.py b/Lib/fnmatch.py
index d5e296f774..28fd843eea 100644
--- a/Lib/fnmatch.py
+++ b/Lib/fnmatch.py
@@ -76,7 +76,11 @@ def translate(pat):
 
     There is no way to quote meta-characters.
     """
+    res = _translate_unwrapped(pat)
+    return fr'(?s:{res})\Z'
+
 
+def _translate_unwrapped(pat):
     STAR = object()
     res = []
     add = res.append
@@ -181,5 +185,4 @@ def translate(pat):
         else:
             add(f"(?>.*?{fixed})")
     assert i == n
-    res = "".join(res)
-    return fr'(?s:{res})\Z'
+    return "".join(res)
```

Reply:
I think it would be best not to use an private function of `fnmatch` from `pathlib`, if possible. That feels like it would create an unfortunate level of coupling between the two modules.

Instinctively I like your other idea of adding an `exclude_prefix` argument to `fnmatch.translate()`; it feels like a clean solution to our problem here. But we shouldn't bloat the `fnmatch` API unnecessarily just to fix our problem here, unless that extra argument is something that end users have actually asked for.

Reply:
I suspect what I want here is fairly niche (unfortunately for me!).

I hesitate to add a `_fnmatch_translate_stripping_prefix` function to pathlib only because it will affect performance slightly IIUC. Perhaps I could add a comment above the `_FNMATCH_SLICE` gubbins to explain what's going on?

Reply:
I wonder if this hack might be more performant than a pass-through function written in Python:

```py
_strip_fnmatch_prefix_and_suffix = operator.itemgetter(slice(_FNMATCH_PREFIX, _FNMATCH_SUFFIX))
```

The `operator` functions are written in C, so that might be pretty fast.

And then lower down, where you currently do `fnmatch.translate(part)[_FNMATCH_SLICE]`, you could do `_strip_fnmatch_prefix_and_suffix(fnmatch.translate(part))` instead.

You'd have to import `operator` at the top of `pathlib`, but I think it may be imported at startup anyway, so it probably wouldn't slow down the import of `pathlib` too much.

Anyway, always worth doing some quick benchmarking! Adding a pass-through Python function here might not actually slow down the code too much; and using `operator` might not speed things up here.

Reply:
Interestingly, using `slice()` without `itemgetter()` seems most performant:

```
$ ./python -m timeit -s 'from fnmatch import translate; pat = translate("*.py"); prefix, suffix = translate("_").split("_")' 'pat[len(prefix):-len(suffix)]'
5000000 loops, best of 5: 86.3 nsec per loop
$ ./python -m timeit -s 'from fnmatch import translate; pat = translate("*.py"); prefix, suffix = translate("_").split("_"); sl = slice(len(prefix), -len(suffix));' 'pat[sl]'
5000000 loops, best of 5: 54.1 nsec per loop
$ ./python -m timeit -s 'from fnmatch import translate; pat = translate("*.py"); from operator import itemgetter; prefix, suffix = translate("_").split("_"); sl = itemgetter(slice(len(prefix), -len(suffix)))' 'sl(pat)'
5000000 loops, best of 5: 62.6 nsec per loop
```

Reply:
I've added a comment giving an explanation for `_FNMATCH_SLICE` - do you think it's sufficient? I think the use of the `slice` object hints towards what's going on too.

Reply:
> I've added a comment giving an explanation for `_FNMATCH_SLICE` - do you think it's sufficient? I think the use of the `slice` object hints towards what's going on too.

Yes, I think that works fine! Won't formally approve as I haven't looked over the PR comprehensively (and don't have the time right now, I have language-summit blogs to finish this evening!!). But the comment looks great, and is very helpful :)

Reply:
Thank you!